### PR TITLE
Cu 86b02r1zr collision calculation

### PR DESCRIPTION
**Pull request by CraftinPark**

### CODE REVIEW

**Comment by CraftinPark** [comment]
Task linked: [CU-86b02r1zr 011 collision calculation](https://app.clickup.com/t/86b02r1zr)

**Comment by CraftinPark** [comment]
uncomment manager.py:68 to see collisions calculation upon car entering manager radius

**Review by changoug — CHANGES_REQUESTED** [review]
Great job! However, it looks like there may be some issues occurring with the algorithm, as two collisions are showing up. Please confirm and address this

File: src/manager/manager.py [hunk 1]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
         if not vehicle_in_list and distance_to_vehicle <= manager.radius: 
             manager.vehicles.append(vehicle)
+            new_vehicle = True
 
         # remove if in list and outside radius
         elif vehicle_in_list and distance_to_vehicle > manager.radius: 
             manager.vehicles.remove(vehicle)
+    return new_vehicle
+
+def get_collisions(manager: Manager, cur_time: float):
```
Comment (changoug) [inline src/manager/manager.py]: add return type

File: src/manager/manager.py [hunk 2]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
+    collisions = []
+    vehicle_pairs = combinations(manager.vehicles, 2)
+    threshold_distance = 2.5 # will be replaced by Alex's defined safety radius
+    
+    for vehicle_pair in vehicle_pairs:
+        vehicle_out_of_bounds_time = int(min(time_until_end_of_route(vehicle_pair[0]), time_until_end_of_route(vehicle_pair[1])))
+        def world_pos_0(t):
```
Comment (changoug) [inline src/manager/manager.py:55]: this is ugly bro. make it pretty. bro.

File: src/manager/manager.py [hunk 3]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
+            return distance(t) - threshold_distance
+        
+        result = minimize_scalar(objective, bounds=(0,vehicle_out_of_bounds_time), method='bounded')
+        if result.success and distance(result.x) <= threshold_distance:
+            time_of_collision = result.x + cur_time
+            # print(f"The objects come within 2.5 meters of each other at t = {time_of_collision}")
+            collisions.append({"vehicle0": vehicle_pair[0], "vehicle1": vehicle_pair[1], "time": time_of_collision})
+    return collisions
+
+def route_position_at_time(vehicle: Vehicle, time: float):
```
Comment (changoug) [inline src/manager/manager.py]: return type

File: src/manager/manager.py [hunk 4]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
+    return vehicle.route_position + vehicle.velocity * time
+
+def time_until_end_of_route(vehicle: Vehicle):
```
Comment (changoug) [inline src/manager/manager.py]: return type

File: src/manager/manager.py [hunk 5]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
+    vehicle_pairs = combinations(manager.vehicles, 2)
+    threshold_distance = 2.5 # will be replaced by Alex's defined safety radius
+    
+    for vehicle_pair in vehicle_pairs:
+        vehicle_out_of_bounds_time = int(min(time_until_end_of_route(vehicle_pair[0]), time_until_end_of_route(vehicle_pair[1])))
+        def world_pos_0(t):
+            return route_position_to_world_position(vehicle_pair[0].route, route_position_at_time(vehicle_pair[0], t))
+        def world_pos_1(t):
+            return route_position_to_world_position(vehicle_pair[1].route, route_position_at_time(vehicle_pair[1], t))
+        def distance(t):
```
Comment (changoug) [inline src/manager/manager.py]: Can be a one-liner once you have the functions for it, or separate distance function

File: src/manager/manager.py [hunk 6]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
+            wp0 = world_pos_0(t)
+            wp1 = world_pos_1(t)
+            return np.linalg.norm(wp1-wp0)
+        def objective(t):
```
Comment (changoug) [inline src/manager/manager.py]: Same as distance function.

File: src/manager/manager.py [hunk 7]
```diff
@@ -62,10 +38,47 @@ def _update_manager_vehicle_list(manager: Manager, vehicles: list[Vehicle]):
...
 
         # remove if in list and outside radius
         elif vehicle_in_list and distance_to_vehicle > manager.radius: 
             manager.vehicles.remove(vehicle)
+    return new_vehicle
+
+def get_collisions(manager: Manager, cur_time: float):
+    collisions = []
+    vehicle_pairs = combinations(manager.vehicles, 2)
+    threshold_distance = 2.5 # will be replaced by Alex's defined safety radius
```
Comment (changoug) [inline src/manager/manager.py]: threshold_distance=2.5 doesn't seem to reflect the current simulation visual. Consider threshold_distance=4, the one we tested.

**Review by changoug — CHANGES_REQUESTED** [review]

File: src/manager/manager.py (hunk 2 above)
Comment (changoug) [inline src/manager/manager.py]: Consider separate function, it is a bit confusing.

**Review by changoug — CHANGES_REQUESTED** [review]

**Review by changoug — APPROVED** [review]
Looks gud

**Pull request merged**